### PR TITLE
feat: prototype Stars Chart for Repo pages

### PR DESCRIPTION
**Pull request by zeucapua**
## Description

Implements a basic bar chart in repo pages that shows the number of stars given to a repository every day based on the chosen range. This PR adds Recharts as an alternative charting library and creates a new function that adds empty days in between those given by the `/histogram/stars` API endpoint.

This only shows the "daily" chart; the "cumulative" chart will be done in a separate PR.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Relates to #3025 

## Mobile & Desktop Screenshots/Recordings

https://github.com/open-sauced/app/assets/20603494/4197a1a8-7fcc-4527-870f-9635f1b422a9

## Steps to QA

1. Go to a repo page
2. Change the range
3. Ensure that tooltips show on hover, even if there's a day with 0 stars.

## Tier (staff will fill in)

- [ ] Tier 1
- [x] Tier 2
- [ ] Tier 3
- [ ] Tier 4

## [optional] What gif best describes this PR or how it makes you feel?

brooklyn-nine-nine-rosa-diaz-gif

### CODE REVIEW

File: lib/hooks/api/useFetchMetricStats.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import useSWR, { Fetcher } from "swr";
+import { publicApiFetcher } from "lib/utils/public-api-fetcher";
+
+type UseFetchMetricStatsProps = {
```
Comment (nickytonline) [inline lib/hooks/api/useFetchMetricStats.ts:4]: refactor (blocking): the type is for params, not props of a component.

```suggestion
type UseFetchMetricStatsParams = {
```

File: lib/hooks/api/useFetchMetricStats.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  const query = new URLSearchParams();
+  query.set("repo", repository);
+  query.set("range", range.toString());
+
+  const endpoint = () => {
+    switch (variant) {
+      case "stars":
+        return `histogram/stars?${query.toString()}`;
+      case "forks":
+        return `histogram/forks?${query.toString()}`;
```
Comment (nickytonline) [inline lib/hooks/api/useFetchMetricStats.ts:26]: refactor (blocking): The template string will convert the `query` object to a string like `query.toString()`.

```suggestion
        return `histogram/stars?${query}`;
      case "forks":
        return `histogram/forks?${query}`;
```

**Comment by bdougie** [comment]
I love it. I think we can continue to iterate after this is merged.

Screen Shot 2024-03-28 at 11 44 14 AM

I did notice the fetching is a little slow, when switching from 30 to 90. It would be nice to have a loading indicator. My instinct was there was 0 stars at first.

File: lib/utils/repo-page-utils.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    previousCount = count;
+    return days;
+  }, {});
+
+  // convert to array
+  const result: StatsType[] = [];
+  for (let i = 0; i < range; i++) {
+    let temp: StatsType | undefined = allDays && allDays[i];
```
Comment (nickytonline) [inline lib/utils/repo-page-utils.ts:37]: refactor (blocking): Use optional chaining. The type is also inferred so unnecessary to explicitly type it.

```suggestion
    let temp = allDays?.[i];
```

**Review by nickytonline — CHANGES_REQUESTED** [review]
This is working great @zeucapua! Some change requests amnd I wonder @isabensusan about a loading state. Between day range changes, the graph was just empty for a few seconds. Some sort of indicator of data loading would be helpful here.

File: components/Graphs/StarsChart.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+type StarsChartProps = {
+  stats: StatsType[] | undefined;
+  syncId: string;
+  range: number;
+};
+
+export default function StarsChart({ stats, syncId, range = 30 }: StarsChartProps) {
+  const chartData = getStarsHistogramToDays({ stats, range });
+  return (
+    <Card className="flex flex-col gap-8 w-full lg:min-w-[64rem] h-full items-center pt-8">
```
Comment (isabensusan) [inline components/Graphs/StarsChart.tsx:18]: Add a title that is composed by the metric's name + the time range currently selected like:

```
<h3 class="font-semibold text-lg text-slate-800 w-full">
   Stars <span class="pl-1 pl-2 text-slate-500 text-sm font-medium"> 90 days</span> 
</h3>
```

**Review by isabensusan — APPROVED** [review]

**Review by nickytonline — APPROVED** [review]

**Pull request merged**